### PR TITLE
Added driver parameter for h5netcdf

### PR DESCRIPTION
**Pull request by zequihg50**
Following [this change in h5netcdf](https://github.com/h5netcdf/h5netcdf/pull/220), allow xarray to open HDF5 files using the ros3 VFD from HDF5.

### CODE REVIEW

**Comment by kmuehlbauer** [comment]
@zequihg50 This is looking good to me. 

I've almost no experience with cloud access data, so please bear with me. It would be great if there is a permanent endpoint somewhere, or if it could be emulated within CI. Not only here, but also over at h5netcdf. Thanks!

**Comment by kmuehlbauer** [comment]
The only thing we might have to think about is, that this will only take effect with the next h5netcdf release. For all other versions this will silently not take `driver` into account and use the current `h5pyd` approach.

**Comment by zequihg50** [comment]
Now that the pull request for h5netcdf is [merged](https://github.com/h5netcdf/h5netcdf/pull/220), is there anything missing here? For testing I have suggested to use a file from netCDF [examples](https://www.unidata.ucar.edu/software/netcdf/examples/OMI-Aura_L2-example.nc).

**Comment by kmuehlbauer** [comment]
> Now that the pull request for h5netcdf is [merged](https://github.com/h5netcdf/h5netcdf/pull/220), is there anything missing here? For testing I have suggested to use a file from netCDF [examples](https://www.unidata.ucar.edu/software/netcdf/examples/OMI-Aura_L2-example.nc).

Yes, a test would be nice indeed. Also, as mentioned over in #8423, it would be nice to distribute additional kwargs to the backend.

File: xarray/backends/h5netcdf_.py [hunk 1]
```diff
@@ -140,6 +140,7 @@ def open(
         invalid_netcdf=None,
         phony_dims=None,
         decode_vlen_strings=True,
+        driver=None,
```
Comment (kmuehlbauer) [inline xarray/backends/h5netcdf_.py:143]: We might want to enable `driver_kwds`, too. See https://docs.h5py.org/en/stable/high/file.html#file-drivers. With that users would be able to correctly specify eg. authentication.

File: xarray/backends/h5netcdf_.py [hunk 2]
```diff
@@ -161,6 +162,7 @@ def open(
         kwargs = {
             "invalid_netcdf": invalid_netcdf,
             "decode_vlen_strings": decode_vlen_strings,
+            "driver": driver,
```
Comment (kmuehlbauer) [inline xarray/backends/h5netcdf_.py:165]: Here we would need to update kwargs, adding `driver_kwds` to `kwargs` if `driver` is given.

File: xarray/backends/h5netcdf_.py [hunk 3]
```diff
@@ -397,6 +399,7 @@ def open_dataset(  # type: ignore[override]  # allow LSP violation, not supporti
         invalid_netcdf=None,
         phony_dims=None,
         decode_vlen_strings=True,
+        driver=None,
```
Comment (kmuehlbauer) [inline xarray/backends/h5netcdf_.py:402]: Here too, add `driver_kwds`.

File: xarray/backends/h5netcdf_.py [hunk 4]
```diff
@@ -407,6 +410,7 @@ def open_dataset(  # type: ignore[override]  # allow LSP violation, not supporti
             invalid_netcdf=invalid_netcdf,
             phony_dims=phony_dims,
             decode_vlen_strings=decode_vlen_strings,
+            driver=driver,
```
Comment (kmuehlbauer) [inline xarray/backends/h5netcdf_.py:413]: Here too, add `driver_kwds`.

**Review by kmuehlbauer — CHANGES_REQUESTED** [review]
@zequihg50 I've added a couple of suggestions. Please let me know, if you can add the needed changes and a test.

Please also add an entry to [whats-new.rst](https://github.com/pydata/xarray/blob/main/doc/whats-new.rst).

File: xarray/backends/h5netcdf_.py (hunk 1 above)
Comment (kmuehlbauer) [reply xarray/backends/h5netcdf_.py:143]:
```suggestion
        driver=None,
        driver_kwds=None,
```

File: xarray/backends/h5netcdf_.py (hunk 3 above)
Comment (kmuehlbauer) [reply xarray/backends/h5netcdf_.py:402]:
```suggestion
        driver=None,
        driver_kwds=None,
```

File: xarray/backends/h5netcdf_.py [hunk 5]
```diff
@@ -161,6 +162,7 @@ def open(
         kwargs = {
             "invalid_netcdf": invalid_netcdf,
             "decode_vlen_strings": decode_vlen_strings,
+            "driver": driver,
         }
```
Comment (kmuehlbauer) [inline xarray/backends/h5netcdf_.py:166]:
```suggestion
        }
        if driver_kwds is not None:
            kwargs.update(driver_kwds)
```

File: xarray/backends/h5netcdf_.py (hunk 4 above)
Comment (kmuehlbauer) [reply xarray/backends/h5netcdf_.py:413]:
```suggestion
            driver=driver,
            driver_kwds=driver_kwds,
```

**Review by kmuehlbauer — COMMENTED** [review]
@zequihg50 I've added some code suggestion how to avoid some issues with CachingFileManager.

**Comment by kmuehlbauer** [comment]
@zequihg50 Thanks for adding the test. The test will only work for h5netcdf >= 1.3.0, though.


@dcherian Do we have to check for h5netcdf-version here? Or does it make sense to just let the test run if h5netcdf >= 1.3.0 is installed?

**Comment by dcherian** [comment]
> Or does it make sense to just let the test run if h5netcdf >= 1.3.0 is installed?

:+1: @zequihg50 we use this type of decorator to run a test only if the version is greater than a min version: https://github.com/pydata/xarray/blob/0c1ad54d67015437b507a33305a30fa2cb1f5398/xarray/tests/__init__.py#L93-L97

**Comment by kmuehlbauer** [comment]
LGTM! @zequihg50 Please merge xarray latest main into this PR branch and add an entry to https://github.com/pydata/xarray/blob/main/doc/whats-new.rst to make this enhancement visible.

**Comment by dcherian** [comment]
Thanks @zequihg50 and @kmuehlbauer

**Comment by kmuehlbauer** [comment]
Thanks @zequihg50 for pushing this through!

**Pull request merged**